scertecd/scertecd.go

```diff
@@ -0,0 +1,862 @@
+	if s.Logf != nil {
+		s.Logf(format, args...)
+	} else {
+		log.Printf(format, args...)
+	}
+}
+
+// UpdateAll checks or updates all certs once and returns.
+//
+// If all certs are either fine or successfully updated, it returns nil.
+//
+// It is not necessary to call Start before UpdateAll.
+func (s *Server) UpdateAll() error {
+	s.initDomainAndTypes()
+	for _, dt := range s.dts {
+		cu := s.newCertUpdateCheck(dt)
+		st, err := cu.updateIfNeeded(context.Background())
+		if err != nil {
+			cu.lg.Printf("updateIfNeeded error: %v", err)
+			return err
+		}
+		cu.lg.Printf("success; %+v", st)
+	}
+	return nil
+}
+
+// Start starts a background renewal goroutine for each cert domain and
+// algorithm type. The context is used only for the initial ACME registration
+// check and not used thereafter.
+func (s *Server) Start(ctx context.Context) error {
+	s.initDomainAndTypes()
+	if _, err := s.getACMEClient(ctx); err != nil {
+		return err
+	}
+	for _, dt := range s.dts {
+		go s.renewCertLoop(dt)
+	}
+	return nil
+}
+
+func (s *Server) getACMEClient(ctx context.Context) (*acme.Client, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.acLazy != nil {
+		return s.acLazy, nil
+	}
+
+	ac, err := s.getOrMakeACMEClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("getOrMakeACMEClient: %w", err)
+	}
+	if err := s.initACMEReg(ctx, ac); err != nil {
+		return nil, fmt.Errorf("initACMEReg: %w", err)
+	}
+	s.acLazy = ac
+	return ac, nil
+}
+
+var tmpls = template.Must(template.New("root").Parse(`
+{{define "root"}}
+<html><h1>scertecd</h1><table border=1 cellpadding=5>
+{{range .Certs}}
+   <tr>
+       <td><b>{{.Name}}</b>
+	   <td>{{.Status}}
+	       {{if .Log}}<pre>{{.Log}}</pre>{{end}}
+	   </td>
+   </tr>
+{{end}}
+</table></html>
+{{end}}
+`))
+
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	type certData struct {
+		Name   string
+		Status string
+		Log    string
+	}
+	var data struct {
+		Certs []certData
+	}
+	now := s.now()
+	addRow := func(cu *certUpdateCheck) {
+		if cu == nil {
+			return
+		}
+		cu.mu.Lock()
+		defer cu.mu.Unlock()
+
+		cd := certData{
+			Name: cu.dt.SecretName(cu.s),
+		}
+
+		if cu.end.IsZero() {
+			cd.Status = "in progress"
+			cd.Log = cu.log.String()
+		} else if cu.err != nil {
+			cd.Status = "error: " + cu.err.Error()
+		} else if cu.res == nil {
+			cd.Status = "unexpected done with no error and no result"
+		} else {
+			cd.Status = fmt.Sprintf("success; version %d; checked %v ago, good for %v",
+				cu.res.SecretVersion,
+				now.Sub(cu.end).Round(time.Second),
+				formatDuration(cu.res.ExpiresAt.Sub(now)))
+		}
+		data.Certs = append(data.Certs, cd)
+	}
+
+	for _, dt := range s.dts {
+		s.mu.Lock()
+		last := s.last[dt]
+		s.mu.Unlock()
+		addRow(last)
+	}
+
+	w.Header().Set("Content-Type", "text/html; charset=utf-8")
+	tmpls.ExecuteTemplate(w, "root", data)
+}
+
+func (s *Server) now() time.Time {
+	if s.Now != nil {
+		return s.Now()
+	}
+	return time.Now()
+}
```

Comment:
similar to `s.Logf`, might be worth to initialize `s.Now = time.Now` if it's nil

Reply:
done